tchannel-core/src/main/java/com/uber/tchannel/handlers/ResponseRouter.java

```diff
@@ -46,18 +56,67 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
         this.ctx = ctx;
     }
 
-    public ListenableFuture<ResponseMessage> expectResponse(RawRequest request) throws InterruptedException {
+    public <V> ListenableFuture<V> expectResponse(Request request) throws InterruptedException {
         int messageId = idGenerator.incrementAndGet();
         request.setId(messageId);
-        SettableFuture<ResponseMessage> future = SettableFuture.create();
+        SettableFuture<V> future = SettableFuture.create();
         this.messageMap.put(request.getId(), future);
+        this.requestMap.put(request.getId(), request);
         ctx.writeAndFlush(request);
         return future;
     }
 
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    protected void messageReceived(ChannelHandlerContext ctx, ResponseMessage response) throws Exception {
+
+        Object future = this.messageMap.remove(response.getId());
+        if (future == null) {
+            return;
+        }
+
+        // TODO: handle timeout here
+        // TODO better interface?
+        Request request = this.requestMap.remove(response.getId());
+        ArgScheme argScheme = request.getArgScheme();
+        if (argScheme == null) {
```

Comment:
argSchemes are as headers in node. It is named this way and I didn't change it. :)

an invalid req can miss as header. We need to handle it.